harness/harness-core
# [feat]: [PL-38975]: Run a migration to fix users with same email but different Harness uuid in CG v/s NG

## Describe your changes
Described in this JIRA comment - https://harness.atlassian.net/browse/PL-38975?focusedCommentId=503456

**Testing:**
- Basic sanity pre and post migration for NG user.
- Tested exact CFD usecase.
     - Add a user in NG and provide them bunch of roles and add them to multiple groups, scopes that have roles themselves -> User will get appropriate access on NG
     - Disable NG in Account and delete and re-create user in CG.
     - Enable NG again
     - At this point, User ids of this user should not match for same email between CG mgr and NG mgr -> User will not be able to open NG now
     - Run my migration. -> User should get all access back

### Pseudo code of what we have implemented:

 for each entry in NG UserMetaData :
     - search with email in CG
         - if not found, then delete entry from UserMetaData.
         - if found, then compare uuid of NG vs CG
             - if uuid same - do nothing 
             - if uuid different
                 - in single transaction {Using mongotemplate - delete userMetaData ng_uuid and add new entry cg_uuid}
                 - fetch all entries from NG:userMembershipV2 for ng_uuid and add corresponding new entries to userMembershipV2 with cg_uuid -> Will create ACLs 
                 - AccessControl:roleassignment
                     - fetch all entries for principalIdentifier=ng_uuid
                     - create corresponding new entries for principalIdentifier=cg_uuid -> This will create ACLs
                     - delete all entries for  principalIdentifier=uuid_old
                 - NG Usergroup update-> delete uuid_old and add uuid_new
                     - This will take care of Access Control's user group
                 - Delete old entry from userMembershipV2 with uuid_old
 

## Checklist

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions


## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/harness/harness-core/48795)
